File: Лабы_2/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int counting_letters(string text);
int counting_words(string text);
int counting_sentence(string text);

int main(void)
{
    // Ввод текста пользователем
    string text = get_string("Text: ");

    // Подсчет кол-ва букв, слов и предложений в тексте
    int letters = counting_letters(text);
    // printf("Кол-во букв: %i\n", letters);

    int words = counting_words(text);
    // printf("Кол-во слов: %i\n", words);

    int sentence = counting_sentence(text);
    // printf("Кол-во предложений: %i\n", sentence);

    float L = (float) letters / (float) words * 100.00;
    // printf("%f\n", L);

    float S = (float) sentence / (float) words * 100.00;
    // printf("%f\n", S);

    // Вычисление индекса Коулмана-Ляу
    float index;
    index = 0.0588 * L - 0.296 * S - 15.8;
    // printf("%f\n", index);

    int grade = round(index);

    if (grade < 1)
    {
        printf("Before Grade 1\n");
    }

    else if (grade > 1 && grade < 17)
    {
        printf("Grade %i\n", grade);
    }
    else
    {
        printf("Grade 16+\n");
    }
}

// Функция расчета кол-ва букв в тексте
int counting_letters(string text)
{
    int letter = 0;

    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if ((text[i] >= 'A' && text[i] <= 'Z') || (text[i] >= 'a' && text[i] <= 'z'))
        {
            letter++;
        }
    }

    return letter;
}

// Функция расчета кол-ва слов в тексте
int counting_words(string text)
{
    int count = 0;
    int inWord = 0; // Флаг, показывающий в слове мы сейчас или нет

    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (isspace(text[i]))
        {
            inWord = 0;
        }
        else if (inWord == 0)
        {
            inWord = 1;
            count++;
        }
    }

    return count;
}

// Функция расчета кол-ва предложений в тексте
int counting_sentence(string text)
{
    int count = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        // Пропускаем пробелы в начале предложения
        while (i < n && isspace(text[i]))
        {
            i++;
        }

        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            if (text[i] == '.' && i + 2 < n && text[i + 1] == '.' && text[i + 2] == '.')
            {
                i += 3;
            }
            count++;
        }

        // Пропускаем пробелы после конца предложения
        while (i + 1 < n && isspace(text[i + 1]))
        {
            i++;
        }
    }

    return count;
}

File: Лабы_6/sentimental-readability/readability.py
from cs50 import get_string
import re

def main():
    text = get_string("Text: ")
    letters = GetLetters(text)
    words = GetWords(text)
    sentences = GetSentence(text)

    # Расчет Среднего кол-ва букв на 100 слов в тексте
    L = letters / words * 100.00
    #print(f"Среднее кол-во букв на 100 слов в тексте: {L}")

    # Расчет Среднего кол-ва предложений на 100 слов в тексте
    S = sentences / words * 100.00
    #print(f"Среднее кол-во предложений на 100 слов в тексте: {S}")

    index = 0.0588 * L - 0.296 * S - 15.8

    # Вывод сообщения об уровне чтения
    if index < 1:
        print("Before Grade 1")
    elif index >= 1 and index <= 16:
        print(f"Grade: {round(index)}")
    elif index > 16:
        print("Grade 16+")

# Функция подсчета букв в тексте
def GetLetters(text):

    # Счетчик букв в тексте
    letters = 0
    for i in text:
        if i.isalpha():
            #print(i)
            letters += 1

    #print(f"Количество букв: {letters}")
    return letters


# Функция подсчета слов в тексте
def GetWords(text):

    # Счетчик слов в тексте
    words = 0
    word = text.split(' ')
    #print(word)
    for i in word:
        #print(i)
        words += 1

    #print(f"Количество слов: {words}")
    return words


# Функция подсчета слов в тексте
def GetSentence(text):

    # Счетчик предложений в тексте
    sentences = 0
    sentence = re.split(r'[.!?...] ', text)
    #print(sentence)
    for i in sentence:
        #print(i)
        sentences += 1

    #print(f"Количество предложений: {sentences}")
    return sentences

main()
